File: self pased DSA/Tree/16-Kth-Largest-Sum-in-a-Binary-Tree.cpp
/*
You are given the root of a binary tree and a positive integer k.
The level sum in the tree is the sum of the values of the nodes that are on the same level.
Return the kth largest level sum in the tree (not necessarily distinct). If there are fewer 
than k levels in the tree, return -1.
Note that two nodes are on the same level if they have the same distance from the root.
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    long long kthLargestLevelSum(TreeNode* root, int k) {
        // we  have to do level order traversel
        vector<long long> levelSum;
        
        queue<TreeNode*> q;
        q.push(root);

        while(!q.empty()){
            int size = q.size();
            // find the sum of all the elements in a perticuler level
            long long sum=0;
            while(size-->0){
                auto node = q.front();
                q.pop();

                sum += node->val;
                // check if its has left child then push that into queue
                if(node->left!=NULL) q.push(node->left);
                // check if its has right child then push that into queue
                if(node->right!=NULL) q.push(node->right);
            }
            levelSum.push_back(sum);
        }

        if(levelSum.size()<k) return -1;
        sort(levelSum.begin(), levelSum.end(), greater<long long>());
        return levelSum[k-1];

    }
};

File: code_java/self pased DSA/Tree/16-Kth-Largest-Sum-in-a-Binary-Tree.java
/**
 * Problem Summary:
 * Given the root of a binary tree and a positive integer k,
 * the level sum of the tree is defined as the sum of all node values at the same depth level.
 * The task is to find the kth largest level sum in the tree.
 * If the tree has fewer than k levels, return -1.
 *
 * Approach:
 * 1. Use a level order traversal (BFS) to traverse the tree level by level.
 * 2. For each level, calculate the sum of node values.
 * 3. Store all level sums in a list.
 * 4. Sort the list of sums in descending order.
 * 5. Return the kth largest sum if it exists, otherwise return -1.
 */

import java.util.*;

public class Solution {
    public long kthLargestLevelSum(TreeNode root, int k) {
        if (root == null) return -1;

        List<Long> levelSum = new ArrayList<>();
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int size = queue.size();
            long sum = 0;

            for (int i = 0; i < size; i++) {
                TreeNode node = queue.poll();
                sum += node.val;

                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }

            levelSum.add(sum);
        }

        if (levelSum.size() < k) return -1;

        // Sort sums in descending order
        Collections.sort(levelSum, Collections.reverseOrder());

        return levelSum.get(k - 1);
    }
}

/**
 * Definition for a binary tree node.
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {}

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}